editor/components/inserter/index.js

```diff
@@ -97,22 +97,29 @@ export default compose( [
 		( state, ownProps ) => {
 			return {
 				insertionPoint: getBlockInsertionPoint( state, ownProps.rootUID ),
+				selectedBlock: getSelectedBlock( state ),
 			};
 		},
-		( dispatch, ownProps ) => ( {
+		{
+			showInsertionPoint,
+			hideInsertionPoint,
+			insertBlock,
+			replaceBlocks,
+		},
+		( { selectedBlock, ...stateProps }, dispatchProps, { layout, rootUID, ...ownProps } ) => ( {
+			...stateProps,
+			...ownProps,
+			showInsertionPoint: dispatchProps.showInsertionPoint,
+			hideInsertionPoint: dispatchProps.hideInsertionPoint,
 			onInsertBlock( item, index ) {
-				const { rootUID, layout } = ownProps;
 				const { name, initialAttributes } = item;
-				dispatch( insertBlock(
-					createBlock( name, { ...initialAttributes, layout } ),
-					index,
-					rootUID,
-				) );
+				const insertedBlock = createBlock( name, { ...initialAttributes, layout } );
+				if ( selectedBlock && isUntouchedDefaultBlock( selectedBlock ) ) {
+					dispatchProps.replaceBlocks( selectedBlock.uid, insertedBlock );
```

Comment:
Random thought: Would this be any less complex if we just allowed the block to be inserted as normal, but had a side-effect to remove the previous untouched default block?

Reply:
What about this feedback?

Reply:
Oh sorry about that.

Are you saying I should call another action `removeBlock` + `insertBlocks` instead of `replaceBlocks`: We'd still need the `selectedBlock` prop to check if it's unmodified or not, so we'd still need `mergeProps`. I'm thinking this doesn't change anything in terms of complexity

Or maybe you're saying we should use `effects`? Maybe it just adds unnecessary indirection don't you think? It will have the advantage of being more generic (though I'm not sure we want this to be generic: inserting at a given position, no matter how we triggered this insert will trigger the behavior)

Reply:
>Or maybe you're saying we should use `effects`?

This is more what I had in mind, that upon inserting a new block, if the block preceding it (a) was previously the last block, (b) is of the default block type, and (c) is empty, then it should be removed.

I guess it could be perceived as indirection. I'd framed it in my mind more as a side effect obviously 😄